// graphdb/lib/repository/add-statement-payload.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }
var StatementPayload = require('../repository/statement-payload');
var XSD = require('../model/types').XSD;

/**
 * Object builder used for constructing a statement addition payload.
 *
 * @class
 * @author Mihail Radkov
 * @author Svilen Velikov
 */
var AddStatementPayload = /*#__PURE__*/function (_StatementPayload) {
  _inherits(AddStatementPayload, _StatementPayload);
  var _super = _createSuper(AddStatementPayload);
  function AddStatementPayload() {
    _classCallCheck(this, AddStatementPayload);
    return _super.apply(this, arguments);
  }
  _createClass(AddStatementPayload, [{
    key: "setLanguage",
    /**
     * Sets the language this statement's object.
     *
     * This makes the statement a literal.
     *
     * @param {string} [language] the object's language
     * @return {AddStatementPayload} the payload for method chaining
     */
    value: function setLanguage(language) {
      this.payload.language = language;
      return this;
    }
    /**
     * Returns the statement object's language.
     *
     * Having a language means it is a literal.
     *
     * @return {string} the language
     */
  }, {
    key: "getLanguage",
    value: function getLanguage() {
      return this.payload.language;
    }
    /**
     * Sets the data type this statement's object.
     *
     * This makes the statement a literal.
     *
     * @param {string} [dataType] the object's data type
     * @return {AddStatementPayload} the payload for method chaining
     */
  }, {
    key: "setDataType",
    value: function setDataType(dataType) {
      this.payload.dataType = dataType;
      return this;
    }
    /**
     * Returns the statement object's data type.
     *
     * Having a data type means it is a literal.
     *
     * @return {string} the data type
     */
  }, {
    key: "getDataType",
    value: function getDataType() {
      return this.payload.dataType;
    }
    /**
     * Sets the statement's object value making it a literal.
     *
     * If the data type is not provided, it will try to autodetect it:
     * <ul>
     *   <li>integer -> xsd:integer</li>
     *   <li>float -> xsd:decimal</li>
     *   <li>boolean -> xsd:boolean</li>
     * </ul>
     * Everything else will be considered as xsd:string.
     *
     * To set a language the data type must be xsd:string.
     *
     * @param {*} value the statements object value
     * @param {string} [type] the statements object data type
     * @param {string} [language] the statements object language
     * @return {AddStatementPayload} the payload for method chaining
     */
  }, {
    key: "setObjectLiteral",
    value: function setObjectLiteral(value, type, language) {
      if (type) {
        if (type === XSD.STRING && language) {
          this.setLanguage(language);
        }
        this.setDataType(type);
      } else {
        this.setDataType(this.detectLiteralType(value));
      }
      this.setObject(value);
      return this;
    }
    /**
     * Autodetect the provided value's data type.
     *
     * @private
     * @param {*} value the value which type will be autodetected
     * @return {string} the detected data type
     */
  }, {
    key: "detectLiteralType",
    value: function detectLiteralType(value) {
      var valueType = _typeof(value);
      if (valueType === 'number') {
        if (value % 1 === 0) {
          return XSD.INTEGER;
        } else {
          return XSD.DECIMAL;
        }
      } else if (valueType === 'boolean') {
        return XSD.BOOLEAN;
      }
      // Default
      return XSD.STRING;
    }
    /**
     * Returns if this statement payload is for a literal. A literal have
     * data type and/or language.
     *
     * @return {boolean} <code>true</code> if it is a literal payload or
     * <code>false</code> otherwise
     */
  }, {
    key: "isLiteral",
    value: function isLiteral() {
      return !!(this.payload.language || this.payload.dataType);
    }
    /**
     * Sets the base URI that is used for resolving any relative URIs in the
     * current payload.
     *
     * @param {string} baseURI the base URI
     * @return {AddStatementPayload} the payload for method chaining
     */
  }, {
    key: "setBaseURI",
    value: function setBaseURI(baseURI) {
      this.payload.baseURI = baseURI;
      return this;
    }
    /**
     * Returns the base URI that is used for resolving any relative URIs.
     *
     * @return {string} the base URI
     */
  }, {
    key: "getBaseURI",
    value: function getBaseURI() {
      return this.payload.baseURI;
    }
  }]);
  return AddStatementPayload;
}(StatementPayload);
module.exports = AddStatementPayload;
